src/services/redisSessionStorage.ts

```diff
@@ -2,64 +2,187 @@
  * This service uses the Redis client to store and retrieve session information.
  */
 
-import { Either, left } from "fp-ts/lib/Either";
+import { Either, left, right } from "fp-ts/lib/Either";
 import * as redis from "redis";
 import { User } from "../types/user";
 import { extractUserFromJson } from "../types/user";
 import { ISessionStorage } from "./iSessionStorage";
+import TokenService from "./tokenService";
+
+export interface ISessionState {
+  readonly expired: boolean;
+  readonly expireAt?: number;
+  readonly newToken?: string;
+  readonly user?: User;
+}
 
 export default class RedisSessionStorage implements ISessionStorage {
   constructor(
     private readonly redisClient: redis.RedisClient,
-    private readonly tokenDuration: number
+    private readonly tokenDurationSecs: number,
+    private readonly tokenService: TokenService
   ) {}
 
   /**
    * {@inheritDoc}
    */
-  public set(token: string, user: User): void {
-    // Set key to hold the string value. This data is set to expire (EX) after
-    // `this.tokenDuration` seconds.
-    // @see https://redis.io/commands/set
-    this.redisClient.set(token, JSON.stringify(user), "EX", this.tokenDuration);
+  public set(
+    token: string,
+    user: User,
+    timestamp: number
+  ): Promise<Either<Error, boolean>> {
+    return new Promise(resolve => {
+      // Set a key to hold the fields value.
+      // @see https://redis.io/commands/hmset
+      this.redisClient.hmset(
+        token,
+        {
+          data: JSON.stringify(user),
+          timestamp
+        },
+        (err, response) => {
+          if (err) {
+            return resolve(left<Error, boolean>(err));
+          }
+
+          resolve(right<Error, boolean>(response));
+        }
+      );
+    });
   }
 
   /**
    * {@inheritDoc}
    */
-  public get(token: string): Promise<Either<Error, User>> {
-    const client = this.redisClient;
+  public get(token: string): Promise<Either<Error, ISessionState>> {
+    return new Promise(resolve => {
+      // Get the fields for this key.
+      // @see https://redis.io/commands/hmget
+      this.redisClient.hgetall(token, (err, value) => {
+        if (err) {
+          // Client returns an error.
+          return resolve(left<Error, ISessionState>(new Error(err.message)));
+        }
+
+        if (value === null || value === undefined) {
+          // Null and undefined values are considered "not found".
+          return resolve(this.sessionNotFound());
+        }
 
+        // Check if the token has expired. We don't remove expired token
+        // because a client can later refresh the session.
+        const timestamp = +value.timestamp;
```

Comment:
why is `+` needed?

Reply:
`value.timestamp` should probably be called `timestampEpochMsec`

Reply:
> why is + needed?
to cast value.timestamp to number (redis stores only strings)

Reply:
and what if it's not a number? :) what you should really do here is to either use `parseInt` or validate `value` against an expected type with io-ts (better)

Reply:
How can I validate value against an expected type with io-ts? The expected type has timestamp defined as number, but the return type from redis is always string.

Reply:
something like this one
https://github.com/gcanti/io-ts-types/blob/master/src/number/NumberFromString.ts

Reply:
ah sorry I misunderstood, then you should use `parseInt`

Reply:
...or what @gunzip is suggesting, that gives you error handling for free

Reply:
In my case `value` is an object with two properties: a string and a number. Redis returns me  an object with two string and I've to validate that the object has that two properties and that the first is a string and the second is a number. How can I do that with io-ts?

Reply:
you cannot do that on `value` since you must also _deserialize_ the values before validating, so you should do it for each property separately:

* `NumberFromStringType` for the `timestamp`
* `JSON.parse` + `UserType` for the other (?)

Reply:
something like:

` ` ` 
const MyObjectType = t.interface({
   value: NumberFromString,
   foobar: t.string
});
type MyObjectType = t.typeOf<typeof MyObjectType>;
const errorOrValidObj = MyObjectType.decode({  value: "1234", foo: "xxx" });
` ` `

Reply:
just want to point out that decode() deserialize recursively; so, as far as the deserialization of NumberFromString is [embedded in the type defintion](https://github.com/gcanti/io-ts-types/blob/master/src/number/NumberFromString.ts#L15) you can deserialize eveything with just one call to decode(), ie:
```
const errorOrDeserializedUser = JSON.parse(value.user)

const MyObjectType = t.interface({
   value: NumberFromString,
   user: User
});
type MyObjectType = t.typeOf<typeof MyObjectType>;
const errorOrValidObj = MyObjectType.decode({  value: "1234", user: errorOrDeserializedUser );
```